File: 0902-minimum-number-of-refueling-stops/0902-minimum-number-of-refueling-stops.py
import heapq
class Solution:
    def minRefuelStops(self, target: int, startFuel: int, stations: List[List[int]]) -> int:
        pq=[]
        result=0
        i=0
        fuel=startFuel
        while fuel<target:
            while i<len(stations) and fuel>=stations[i][0]:
                heapq.heappush(pq,-stations[i][1])
                i+=1
            if len(pq)==0:
                return -1
            result+=1
            fuel+=-heapq.heappop(pq)
        return result

        

File: 0902-minimum-number-of-refueling-stops/0902-minimum-number-of-refueling-stops.java
class Solution {
    public int minRefuelStops(int target, int startFuel, int[][] stations) {
        PriorityQueue<int[]> pq= new PriorityQueue<>((a,b)->b[1]-a[1]);
        int result=0,j=0;
        while(startFuel<target){
            while(j<stations.length && startFuel>=stations[j][0]){
                pq.add(stations[j]);
                j++;
            }
            if(pq.isEmpty()) return -1;
            startFuel+=pq.poll()[1];
            result++;
        }
        return result;
    }
}